### DISCUSSION TITLE
Decisión de diseño

### DISCUSSION THREAD

**Discussion by Franpanozzo** [dds-jv/foro#306] [Consultas de objetos] [answered]
Buenas, tengo una duda sobre una decisión de diseño. Si tengo una lista de objetos que respetan una interfaz, si yo quiero hacer una cosa para los que son de cierto tipo (de cierta clase que implementa esa interfaz) , me conviene:
1) Mandarle el mismo mensaje a todos y con las clases que no quiero hacer nada que tengan un método vacío. Ejemplo:

```
public class RepositorioConsultas {
  List<Consulta> consultas = new ArrayList<>();

  public void chequearFinalizacion() {
    consultas.forEach(Consulta::chequearFinalizacion);
  }
}

public abstract class Consulta {
    public abstract void chequearFinalizacion();
}

public class Encuesta extends Consulta {
  @Override
  public void chequearFinalizacion() {
     //Hacer algo especifico de requerimiento
  }

public class Formularios extends Consulta{
  @Override
  public void chequearFinalizacion() {

  }
```

2) Tener como una especie de type test y después castear al tipo que yo quiero para poder mandar mensajes de esa
clase ya que no son polimórficos esos mensajes porque son mensajes específicos para el requerimiento de esa clase, ahorrándome el método vacío en los Formularios. Ejemplo:

```
public class RepositorioConsultas {
  List<Consulta> consultas = new ArrayList<>();

  public void chequearFinalizacion() {
        consultas.forEach(consulta -> this.hacerParaEncuesta(consulta));
  }

  public void hacerParaEncuesta(Consulta consulta) {
    if(consulta.esEncuesta) {  
        Encuesta encuesta = (Encuesta) consulta;
       //Hacer cosas especificas para la encuesta
    }
   //Hay un método abstracto en la super clase que implementan después las subclases diciéndonos si son encuesta

   //Podria ser que primero se filtre por las que son encuestas y luego hacer lo que quiera, pero seria lo mismo de castear para 
  //poder mandar estos mensajes especificos que no estan en la super clase, es otra forma.
 }

}
```

Por el momento me decidí por la primera, si me dicen cual es mejor buenísimo, Saludos!

**Comment by roli93** [comment, marked as answer]
Buenas! 
Perdón por el delay, pero, como dice el refrán, mejor tarde que nunca.

Respecto de lo que consultas:

Respuesta corta: Opción 1, en principio no tiene nada de malo tener métodos vacíos para garantizar polimorfismo si realmente, para esa acción, el comportamiento de determinadas subclases es "no hacer nada".

Respuesta no tan corta: Siempre hay que ver la situación, si bien en general no hay problema con tener métodos vacíos, hay que analizar que no haya un problema en otro lado que lo esté causando. Por ej, si tenes 4 subclases, 3 no hacen nada y una sí, es muy probable que haya algo que no está del todo bien (por ej, podrías plantearte si ese método realmente debe ser polimórfico).
En el caso que planteas en general podría estar bien si es que realmente los formularios no tienen que chequear nada y simplemente pueden considerarse finalizados siempre. Sin embargo, lo que sí me hace ruido de ese diseño es que el metodo sea void. Daría la impresión de que lo que realmente queres es _preguntar_ si la consulta está finalizada. En ese caso sería muy simple ver que la Encuesta haría algo, y luego retornaría el valor correspondiente, mientras que el Formulario simplemente devovlería siempre `true`. Me intriga saber qué hace ese método en el caso de la encuesta,si muta una variable o algo... lo cual podría ser un problema en el diseño que -colateralmente- te esté causando este dilema.
